File: 42. Trapping Rain Water/trap.py
from typing import List

class Solution:
  def trap(self, height: List[int]) -> int:
    n = len(height)
    leftMax = [0] * n
    rightMax = [0] * n

    leftMax[0] = 0
    for i in range(1, n):
      leftMax[i] = max(leftMax[i - 1], height[i - 1])

    rightMax[n - 1] = 0
    for i in range(n - 2, - 1, -1):
      rightMax[i] = max(rightMax[i + 1], height[i + 1])

    waterTrapped = 0
    for i in range(n):
      waterAboveBar = min(leftMax[i], rightMax[i]) - height[i]
      waterTrapped += max(waterAboveBar, 0)

    return waterTrapped

solution = Solution()
height = [0,1,0,2,1,0,1,3,2,1,2,1]
print('trap: ', solution.trap(height))

File: 42. Trapping Rain Water/trap.js
/**
 * @param {number[]} height
 * @return {number}
 */
const trap = (height) => {
  if (!height || !height.length) {
    return 0;
  }

  const n = height.length;
  const leftMax = Array(n).fill(0);
  const rightMax = Array(n).fill(0);

  for (let i  = 1; i < n; i++) {
    leftMax[i] = Math.max(leftMax[i - 1], height[i - 1]);
  }

  for (let j = n - 2; j >= 0; j--) {
    rightMax[j] = Math.max(rightMax[j + 1], height[j + 1]);
  }

  let waterTrapped = 0;
  for (let i  = 0; i < n; i++) {
    let waterAboveBar = Math.min( leftMax[i], rightMax[i]) - height[i];
    waterTrapped += Math.max(waterAboveBar, 0);
  }

  return waterTrapped;
};

height = [0,1,0,2,1,0,1,3,2,1,2,1]
console.log('trap: ', trap(height))
